sled-agent/src/storage_manager.rs

```diff
@@ -1203,17 +1211,49 @@ impl StorageManager {
 
     /// Adds a zpool to the storage manager.
     // Receiver implemented by [StorageWorker::upsert_zpool].
-    pub async fn upsert_zpool(&self, name: ZpoolName) {
+    pub async fn upsert_zpool(
+        &self,
+        name: ZpoolName,
+        disk_path: Option<PathBuf>,
+    ) {
         info!(self.log, "Inserting zpool: {name:?}");
-        self.tx.send(StorageWorkerRequest::NewPool(name)).await.unwrap();
+        self.tx
+            .send(StorageWorkerRequest::NewPool { pool_name: name, disk_path })
+            .await
+            .unwrap();
     }
 
     pub async fn get_zpools(&self) -> Result<Vec<crate::params::Zpool>, Error> {
+        let disks = self.resources.disks.lock().await;
         let pools = self.resources.pools.lock().await;
-        Ok(pools
-            .keys()
-            .map(|zpool| crate::params::Zpool { id: zpool.id() })
-            .collect())
+
+        let mut zpools = Vec::with_capacity(pools.len());
+
+        for (name, pool) in pools.iter() {
+            let disk_type = if let Some(disk_path) = &pool.disk_path {
+                if let Some(disk) = disks.get(disk_path) {
+                    disk.variant().into()
+                } else {
+                    // If the zpool claims to be attached to a disk that we
+                    // don't know about, that's an error.
+                    return Err(Error::ZpoolNotFound(
```

Comment:
If we bail here, what happens?   Will the sled agent be able to continue working?

Reply:
Yes - this is only an error being returned from a querying function, which is only called by RSS surveying the zpool landscape right now.

I could opt to simply ignore the zpool causing the error, but I'm leaning towards "loud errors, returned as early as possible" while we're still in development.

Reply:
I think being loud about errors is good.
I just was concerned about the rest of sled agent making forward progress if errors happen.